### PR TITLE
Using Utimes instead of utime hence setting millisecond attribute of last access and last modified time to non-zero

### PR DESCRIPTION
**Pull request by Anipik**
utimensat documentation http://man7.org/linux/man-pages/man2/utimensat.2.html
Fixes https://github.com/dotnet/corefx/issues/31379

### CODE REVIEW

**Comment by Anipik** [comment]
@dotnet-bot test OSX x64 Debug Build

File: src/Common/src/Interop/Unix/System.Native/Interop.UTimensat.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ internal struct UTimBuf
         /// <returns>
         /// Returns 0 on success; otherwise, returns -1 
         /// </returns>
-        [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_UTime", SetLastError = true)]
-        internal static extern int UTime(string path, ref UTimBuf time);
+        [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_UTimensat", SetLastError = true)]
+        internal static extern int UTimensat(string path, TimeSpec[] times);
```
Comment (safern) [inline src/Common/src/Interop/Unix/System.Native/Interop.UTimensat.cs]: Why not passing this as a pointer instead?

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 2]
```diff
@@ -199,11 +199,22 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
...
+            buf[0].TvSec = accessTime ?? _fileStatus.ATime;
+            buf[1].TvSec = writeTime ?? _fileStatus.MTime;
+
+            // setting microsecond part
+            if (accessTime == null)
+                buf[0].TvNsec = _fileStatus.ATimeNsec;
+            if (writeTime == null)
+                buf[1].TvNsec = _fileStatus.MTimeNsec;
+
+            Interop.CheckIo(Interop.Sys.UTimensat(Path.GetFullPath(path), buf), path, InitiallyDirectory);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: Why did you add GetFullPath here? Shouldn't you be using AT_FDCWD?

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 3]
```diff
@@ -199,11 +201,20 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
...
             _fileStatusInitialized = -1;
             EnsureStatInitialized(path);
-            Interop.Sys.UTimBuf buf;
-            // we use utime() not utimensat() so we drop the subsecond part
-            buf.AcTime = accessTime ?? _fileStatus.ATime;
-            buf.ModTime = writeTime ?? _fileStatus.MTime;
-            Interop.CheckIo(Interop.Sys.UTime(path, ref buf), path, InitiallyDirectory);
+
+            // we use utimes() to set the accessTime and writeTime
+            Span<Interop.Sys.TimeSpec> buf = stackalloc Interop.Sys.TimeSpec[2];;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: nit: extra `;`

File: src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs [hunk 4]
```diff
@@ -77,7 +77,7 @@ public void CanGetAllTimesAfterCreation()
         }
 
         [ConditionalFact(nameof(isNotHFS))] // OSX HFS driver format does not support millisec granularity
-        public void TimesIncludeMillisecondPart_Unix()
+        public void TimesIncludeMillisecondPart()
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: Why are you changing the names of these methods? Shouldn't this file be reverted?

Comment (Anipik) [reply src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: Earlier there was 2 different tests for windows and unix. But i merged them together in https://github.com/dotnet/corefx/pull/31164 
So now this test runs on windows as well as unix too

File: src/Native/Unix/System.Native/pal_time.h [hunk 5]
```diff
@@ -6,19 +6,20 @@
 
 #include "pal_compiler.h"
 #include "pal_types.h"
+#include <time.h>
 
-typedef struct UTimBuf
+typedef struct TimeSpec
```
Comment (JeremyKuhne) [inline src/Native/Unix/System.Native/pal_time.h]: There must be a reason they didn't use `time_t` directly in the old struct. I see that it's type is unspecified in the C standard. Does it ever really change? It either does and you need to change back to int64_t, or it doesn't and you can map it directly (eliminate the struct).  Additionally, `long` in C++ is `int` in C#.

File: src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: Ok, thanks

File: src/Native/Unix/System.Native/pal_time.h (hunk 5 above)
Comment (Anipik) [reply src/Native/Unix/System.Native/pal_time.h]: >  Does it ever really change ?

It may change in future
```
The fact that it is in seconds is stated by the POSIX specification, so, if you're coding for POSIX-compliant environments, you can rely on that.
The C++ standard also states that time_t must be an arithmetic type.
Anyway, the unix timing system (second since the Epoch) is going to overflow in 2038.
 So, it's very likely that, before this date, C++ implementations will switch to other non-int datatypes (either a 64b int or a more complex datatype). 
Anyway, switching to a 64b int would break binary compatibility with previous code (since it requires bigger variables), and everything should be recompiled. 
Using 32b opaque handles would not break binary compatibility, 
you can change the underlying library, and everything still works, but time_t would not a time in seconds anymore, 
it'd be an index for an array of times in seconds. 
For this reason, it's suggested that you use the functions you mentioned to manipulate time_t values, and do not assume anything on time_t.
```
I will revert back to using to using int64

> Additionally, long in C++ is int in C#.

i tried this earlier also and again to0. Converting it to int is leading to failure of all the tests in 
https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs#L45 
i am not able to find out why

Comment (JeremyKuhne) [reply src/Native/Unix/System.Native/pal_time.h]: > Converting it to int is leading to failure 

Sorry, long is int with the VC compiler. I think it depends on the target architecture with Unix. Use int64_t to always get 64 bits.

**Comment by Anipik** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by Anipik** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by Anipik** [comment]
@dotnet-bot test OSX x64 Debug Build

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 6]
```diff
@@ -199,11 +201,20 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
             // force a refresh so that we have an up-to-date times for values not being overwritten
             _fileStatusInitialized = -1;
             EnsureStatInitialized(path);
-            Interop.Sys.UTimBuf buf;
-            // we use utime() not utimensat() so we drop the subsecond part
-            buf.AcTime = accessTime ?? _fileStatus.ATime;
-            buf.ModTime = writeTime ?? _fileStatus.MTime;
-            Interop.CheckIo(Interop.Sys.UTime(path, ref buf), path, InitiallyDirectory);
+
+            // we use utimes() to set the accessTime and writeTime
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: nit `utimes()/utimensat()`

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 7]
```diff
@@ -199,11 +201,20 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
...
+            Span<Interop.Sys.TimeSpec> buf = stackalloc Interop.Sys.TimeSpec[2];
+
+            // setting second part
+            buf[0].TvSec = accessTime ?? _fileStatus.ATime;
+            buf[1].TvSec = writeTime ?? _fileStatus.MTime;
+
+            // setting microsecond part
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: Nit, 'nanosecond'

File: src/System.IO.FileSystem/tests/File/GetSetTimes.cs [hunk 8]
```diff
@@ -92,5 +92,20 @@ public void PageFileHasTimes()
...
+        {
+            string firstFile = GetTestFilePath();
+            string secondFile = GetTestFilePath();
+
+            File.WriteAllText(firstFile, "");
+            File.WriteAllText(secondFile, "");
+
+            File.SetLastAccessTimeUtc(secondFile, DateTime.UtcNow);
+            long firstFileTicks = File.GetLastWriteTimeUtc(firstFile).Ticks;
+            long secondFileTicks = File.GetLastWriteTimeUtc(secondFile).Ticks;
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: @anipik do you need a test that verifies that after SetLastWriteTime, the milliseconds are non zero (999/1000 times anyway). Presumably existing tests aren't enough since they were passing without this fix.

File: src/System.IO.FileSystem/tests/File/GetSetTimes.cs [hunk 9]
```diff
@@ -92,5 +92,20 @@ public void PageFileHasTimes()
...
+        public void SetLastWriteTimeMilliSec()
+        {
+            string firstFile = GetTestFilePath();
+            string secondFile = GetTestFilePath();
+
+            File.WriteAllText(firstFile, "");
+            File.WriteAllText(secondFile, "");
+
+            File.SetLastAccessTimeUtc(secondFile, DateTime.UtcNow);
+            long firstFileTicks = File.GetLastWriteTimeUtc(firstFile).Ticks;
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: @Anipik since we're now bringing a second fix to milliseconds timestamps to ship room they will all whether we have checked the test of the BCL to make sure there isn't a third. If you haven't already I would look at all places we directly handle file time stamps in our native and managed code to make sure we are done with these problems.

Comment (Anipik) [reply src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: I have taken a look at other places. But i will test all the properties independently for millisec/nanosecond to be sure on unix systems.

File: src/System.IO.FileSystem/tests/File/GetSetTimes.cs (hunk 8 above)
Comment (Anipik) [reply src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: The current tests doesnt check the millisecond attribute for any operation, i think the problem lies here https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs#L51
We are just setting seconds(0) and milliseconds will automatically take value of zero. 
I suggest we just change this to some non-zero value for both seconds and milliseconds.
We can even check ticks in all cases as that will check the accuracy of operations upto 100 NanoSeconds.

File: src/System.IO.FileSystem/tests/File/GetSetTimes.cs (hunk 9 above)
Comment (Anipik) [reply src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: @danmosemsft [update] there are still some places\codePaths where we need to correct things. I will put up the PR soon for that

Comment (danmosemsft) [reply src/System.IO.FileSystem/tests/File/GetSetTimes.cs]: Great, lets get all ducks in a row and PR(s) in master etc before involving shiproom

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 10]
```diff
@@ -199,11 +201,20 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
...
+
+            // setting second part
+            buf[0].TvSec = accessTime ?? _fileStatus.ATime;
+            buf[1].TvSec = writeTime ?? _fileStatus.MTime;
+
+            // setting nanosecond part
+            buf[0].TvNsec = accessTime == null ? _fileStatus.ATimeNsec : 0;
+            buf[1].TvNsec = writeTime == null ? _fileStatus.MTimeNsec : 0;
+
+            Interop.CheckIo(Interop.Sys.UTimensat(path, ref MemoryMarshal.GetReference(buf)), path, InitiallyDirectory);
```
Comment (stephentoub) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: I don't think the use of Span really buys much here.  Seems like it's attempt to avoid using unsafe code, but the UTimensat marshaling is effectively already unsafe, and the use of ref here is going to incur higher marshaling costs.  I think it'd be better to just stackalloc into a pointer, and have the interop signature continue to take a pointer rather than a ref.

File: src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs [hunk 11]
```diff
@@ -199,11 +201,20 @@ private void SetAccessWriteTimes(string path, long? accessTime, long? writeTime)
...
+            // we use utimes()/utimensat() to set the accessTime and writeTime
+            Span<Interop.Sys.TimeSpec> buf = stackalloc Interop.Sys.TimeSpec[2];
+
+            // setting second part
+            buf[0].TvSec = accessTime ?? _fileStatus.ATime;
+            buf[1].TvSec = writeTime ?? _fileStatus.MTime;
+
+            // setting nanosecond part
+            buf[0].TvNsec = accessTime == null ? _fileStatus.ATimeNsec : 0;
+            buf[1].TvNsec = writeTime == null ? _fileStatus.MTimeNsec : 0;
```
Comment (alexischr) [inline src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: @Anipik I am attempting to backport this to Mono, so I am looking over this. I don't understand here how `.TvSec` and `.TvNSec` are both set to the same value (the `accessTime long` or the `writeTime long`).Before seeing this was merged for CoreFX 3.0,  I felt I had to change `SetAccessWriteTimes` to provide four values instead of two for subsecond resolution (please see my PR on the Mono CoreFX fork, which uses `timeval/utimes()` but should otherwise be similar: https://github.com/mono/corefx/pull/241/files)

Comment (Anipik) [reply src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: the key point here is that we will be either setting setting the lastwritetime or lastaccesstime which implies in the buf struct , one of the two elements will be zero and otherone will be set to the required value

Comment (alexischr) [reply src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: Ah, this looks to be corrected in https://github.com/dotnet/corefx/pull/31593

Comment (alexischr) [reply src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: Yeah but isn't this always using the subsecond part that's from `_fileStatus` (or 0)? So the users of this method (`SetLastAccessTime` and `SetLastWriteTime`) still can't pass-down the subsecond part of the `DateTimeOffset` they receive

Again, this appears fixed in the subsequent PR

Comment (Anipik) [reply src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: i just looked at the most recent version https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

Sry i deleted the previous comment.
So as i mentioned at one time we are setting only one of them(access or write), so we can just set the other one to the current value.


> So the users of this method (SetLastAccessTime and SetLastWriteTime) still can't pass-down the subsecond part of the DateTimeOffset they receive

The accuracy will be upto 100 nanoseconds, this is due to the limitations of DateTime infrastructure

Comment (alexischr) [reply src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs]: If there is an invocation `SetAccessWriteTimes(path, some_accesstime, null)` 
then:
`buf[0].TvSec = some_accessTime`
`buf[0].TvNSec = 0` 
`buf[1].TvSec = _fileStatus.MTime`
`buf[1].TvNSec = _fileStatus.MTimeNSec`

So this patch *preserves* the nanosecond part of the timestamp that's not being changed, but still can't properly set it for the timestamp that is being changed. 

The version on `master` appears to be correct, this is just a backporting concern.
